cpp/src/sampling/detail/prepare_next_frontier_impl.cuh

```diff
@@ -0,0 +1,209 @@
+  vertex_partition_view_t<vertex_t, multi_gpu> vertex_partition,
+  std::vector<vertex_t> const& vertex_partition_range_lasts,
+  bool carry_over_sources,
+  bool dedupe_sources,
+  bool do_expensive_check)
+{
+  vertex_partition_device_view_t<vertex_t, multi_gpu> d_vertex_partition(vertex_partition);
+
+  if (vertex_used_as_source) {
+    auto& [verts, labels] = *vertex_used_as_source;
+
+    size_t current_size = verts.size();
+    verts.resize(current_size + starting_vertices.size(), handle.get_stream());
+    raft::copy(verts.begin() + current_size,
+               starting_vertices.data(),
+               starting_vertices.size(),
+               handle.get_stream());
+
+    if (starting_vertex_labels) {
+      labels->resize(current_size + starting_vertex_labels->size(), handle.get_stream());
+
+      raft::copy(labels->begin() + current_size,
+                 starting_vertex_labels->data(),
+                 starting_vertex_labels->size(),
+                 handle.get_stream());
+
+      auto begin_iter = thrust::make_zip_iterator(verts.begin(), labels->begin());
+
+      thrust::sort(handle.get_thrust_policy(), begin_iter, begin_iter + verts.size());
+
+      auto new_end =
+        thrust::unique(handle.get_thrust_policy(), begin_iter, begin_iter + verts.size());
+
+      verts.resize(thrust::distance(begin_iter, new_end), handle.get_stream());
+      labels->resize(thrust::distance(begin_iter, new_end), handle.get_stream());
+    } else {
+      thrust::sort(handle.get_thrust_policy(), verts.begin(), verts.end());
+
+      auto new_end = thrust::unique(handle.get_thrust_policy(), verts.begin(), verts.end());
+
+      verts.resize(thrust::distance(verts.begin(), new_end), handle.get_stream());
+    }
+  }
+
+  size_t frontier_size = new_frontier_vertices.size();
+  if (carry_over_sources) { frontier_size += starting_vertices.size(); }
+
+  rmm::device_uvector<vertex_t> frontier_vertices(frontier_size, handle.get_stream());
+  auto frontier_vertex_labels =
+    new_frontier_vertex_labels
+      ? std::make_optional<rmm::device_uvector<label_t>>(frontier_size, handle.get_stream())
+      : std::nullopt;
+
+  if (carry_over_sources) {
+    raft::copy(frontier_vertices.begin(),
+               starting_vertices.data(),
+               starting_vertices.size(),
+               handle.get_stream());
+
+    raft::copy(frontier_vertices.begin() + starting_vertices.size(),
+               new_frontier_vertices.data(),
+               new_frontier_vertices.size(),
+               handle.get_stream());
+  } else {
+    raft::copy(frontier_vertices.begin(),
+               new_frontier_vertices.data(),
+               new_frontier_vertices.size(),
+               handle.get_stream());
+  }
+
+  if (frontier_vertex_labels) {
+    if (carry_over_sources) {
+      raft::copy(frontier_vertex_labels->begin(),
+                 starting_vertex_labels->data(),
+                 starting_vertex_labels->size(),
+                 handle.get_stream());
```

Comment:
Same here. Duplicating the code for first copy can be avoided.

Reply:
Fixed in next push